2147 generate iRODS tickets

alvacouch opened the pull request:
This is the second part of 2147. It contains the basic REST API for generating tickets. Note that write tickets are supported but commented out because I have not written the metadata handlers for them.

pkdash commented:
@hyi Can you review this? I am not familiar with the iRODS ticketing system.

alvacouch commented:
@hyi @aphelionz This needs review in order to move forward. When can you get to it?

pkdash commented on hs_core/irods.py:
```diff
@@ -94,11 +102,11 @@ def create_ticket(self, user, path=None, write=False, allowed_uses=1):
                                        .format(user.username, self.short_id))
         else:
-            if not user.uaccess.can_view_resource(self):
+            if not self.raccess.public and not user.uaccess.can_view_resource(self):
```
For unauthorized user ```user.uaccess``` would raise exception.

alvacouch replied:
Fixed. See revision using is_authenticated()

pkdash commented on hs_core/views/resource_ticket_rest_api.py:
```diff
@@ -0,0 +1,194 @@
+
+    REST URL: hsapi/resource/{pk}/{op}/{path}/
+    HTTP methods: GET
+    Returns HTTP 400, 403, 404
```
I think the return should be documented as what the json data that would be returned on success.

pkdash commented on hs_core/views/resource_ticket_rest_api.py:
```diff
@@ -0,0 +1,194 @@
+
+        try:
+            if write:
+                resource, authorized, user = view_utils.authorize(
```
May be better to set a variable like this and then call ```authorize()```:
```needed_permission = ACTION_TO_AUTHORIZE.EDIT_RESOURCE if write else ACTION_TO_AUTHORIZE.VIEW_RESOURCE```

pkdash commented on hs_core/views/resource_ticket_rest_api.py:
```diff
@@ -0,0 +1,194 @@
+        else:  # op == 'write'
+            return Response("Write operation must specify path", status=status.HTTP_400_BAD_REQUEST)
+
+        # if not resource.supports_folders:
```
Better to remove these commented code lines.

pkdash commented on hs_core/views/resource_ticket_rest_api.py:
```diff
@@ -0,0 +1,194 @@
+
+    REST URL: hsapi/resource/{pk}/bag/
+    HTTP methods: GET
+    Returns HTTP 400, 403, 404
```
Same comment as the previous endpoint.

pkdash commented:
@alvacouch I have left few comments for your consideration. It would be good for someone else to review this PR too.

pkdash commented:
@hyi Can you review this soon? This PR has been out there for quite a while.

hyi commented on hs_core/irods.py:
```diff
@@ -14,10 +14,18 @@ class ResourceIRODSMixin(models.Model):
+    @property
+    def irods_home_path(self):
         """ Return the home path for local iRODS resources """
         return settings.IRODS_CWD
```
@alvacouch is this ```irods_home_path``` ever used outside this class? I am just wondering whether there is a need to make this a property of the class. ```settings.IORDS_CWD``` appears pretty straightforward to me and can be directly used.

alvacouch replied:
The reason for this so that someone doesn't have to remember that settings.IRODS_CWD is needed. This documents that. The reason that it is public is that this is required in order to implement this code as a mixin; private code would not be seen in hs_core/models.py:AbstractResource.

hyi commented on hs_core/irods.py:
```diff
@@ -94,11 +102,11 @@ def create_ticket(self, user, path=None, write=False, allowed_uses=1):
         # authorize user
         if write:
-            if not user.uaccess.can_change_resource(self):
+            if user.username == 'anonymous' or not user.uaccess.can_change_resource(self):
```
```user.is_authenticated" can be used to check for anonymous user or not

alvacouch replied:
Done.

hyi commented on hs_core/irods.py:
```diff
@@ -140,7 +148,8 @@ def create_ticket(self, user, path=None, write=False, allowed_uses=1):
                              'expires', formatted)
-        return ticket
+
+        return ticket, self.irods_full_path(path)
```
@alvacouch Since line 132 already formed the full path to assign to ```path```, it seems ```path``` can be directly returned without calling ```self.irods_full_path(path)```

alvacouch replied:
It fully qualifies local paths with the local prefix. It is not a no-op.

hyi replied:
It seems to me that the line 132 as shown below does the same thing as calling ```self.irods_full_path(path)```, hence my comment above.
```
if path.startswith(self.short_id) or path.startswith('bags/'):  # local path
            path = os.path.join(self.irods_home_path, path)
```

hyi commented on hs_core/irods.py:
```diff
@@ -140,7 +148,8 @@ def create_ticket(self, user, path=None, write=False, allowed_uses=1):
         formatted = timeout.strftime("%Y-%m-%d.%H:%M")
         istorage.session.run('iticket', None, 'mod', ticket,
                              'expires', formatted)
```
@alvacouch The iticket mod command parameter appears to be ```expire``` rather than ```expires``` according to iRODS documentation https://docs.irods.org/4.1.3/icommands/tickets/. This is important to check to make sure, since otherwise, the ticket will not expire as specified, which would create security issues.

alvacouch replied:
Done.

hyi commented on hs_core/irods.py:
```diff
@@ -156,27 +165,28 @@ def list_ticket(self, ticket):
-                            output['fed_path'] = self.resource_federation_path
+                            output['home_path'] = self.resource_federation_path
                         else:
                             location = value.search(self.short_id)
```
@alvacouch Are you sure ```search()``` would work here? ```value``` appears to be a string, and there does not appear to be a ```search``` method for string, do you mean to use ```find``` instead?

alvacouch replied:
Correct. Find it is.

hyi commented on hs_core/irods.py:
```diff
@@ -156,27 +165,28 @@ def list_ticket(self, ticket):
+                                output['home_path'] = self.irods_home_path
+                            else:
+                                output['long_path'] = value[location:]
+                                output['home_path'] = value[:(location-1)]
```
@alvacouch I believe it should be ```output['home_path'] = value[:location]```

alvacouch replied:
The -1 omits a trailing slash. Otherwise it would be included.

hyi replied:
It is better documented then and I assume the trailing slash will always be present from ticket listing output.

alvacouch replied:
I will document this in the code.

hyi commented:
@alvacouch I am done with the review and have left some comments for you to address or clarify.

hyi commented:
@alvacouch I have a couple of further comments for your consideration. Otherwise, a +1 from me.

alvacouch commented:
@hyi Currently there is no handling of whether the resource can contain folders or not. 

* At present, with only `read` tickets being generated, this is not an issue. If the folder exists and shouldn't, one will still be able to read it. This is not particularly bad. 
* In the future, with `write` tickets, I think it will become an issue. We should prevent ticket generation for folders that should not exist.

hyi commented:
@alvacouch That makes sense to me. To take it a bit further, not only it is needed for checking whether the folder creation is allowed or not for the resource for a write ticket, but perhaps it is also needed to check the correct content files are written for the resource, e.g., only .zip file and .tiff file, and .vrt files are allowed for raster resources, etc.. So a more complete specific resource type file content validation function may be needed for this. I suggest this more elaborate comments be put into here for future reference if you agree.

alvacouch commented:
@hyi I added more comments that will probably be enough for now. I renamed the field 'ticket' to 'ticket_id' in responses. 

I am going to wait to merge this until I have a client running that validates the approach.

hyi commented:
@alvacouch Sounds good.

alvacouch commented:
@hyi @aphelionz I have more work to do on the client, But @Castronova wants the ticket interface merged now for use with JupyterHub. This would enable read-only access to resources via access control. As it already works -- and just lacks a client interface -- I think it's time to merge it. 

Do you agree?

hydrobuild commented:
Build finished.

aphelionz commented:
@alvacouch I agree! but testing did not succeed :(

alvacouch commented:
@aphelionz You moved views to core_views in hs_rest_api/urls.py and I merged my changes naively. Fixed.

hydrobuild commented:
Build finished.

aphelionz commented:
Great. @pkdash Can you review? Who else should look at this?

alvacouch commented:
@aphelionz: @hyi already reviewed this. She didn't ever approve it. It is unchanged since the last interchange.

aphelionz commented:
@hyi Please formally approve if you still approve

aphelionz commented:
@alvacouch Go ahead and merge when ready
